### PR TITLE
Handling null arguments in ssl.c

### PR DESCRIPTION
**Pull request by nmittler**
Motivation:

There are several methods in ssl.c that do not verify the parameters before using them, which has been the cause of segfaults with openssl variants (i.e. BoringSSL).

Modifications:

Tightening up the API methods in ssl.c to properly handle null arguments.

Result:

ssl.c should do better at guarding against segfaults

### CODE REVIEW

**Comment by nmittler** [comment]
@normanmaurer @Scottmitch PTAL

File: src/main/c/ssl.c [hunk 1]
```diff
@@ -1618,6 +1753,7 @@ TCN_IMPLEMENT_CALL(jlong, SSL, getTime)(TCN_STDARGS, jlong ssl)
     if (ssl_->session == NULL) {
         // BoringSSL does not protect against a NULL session. OpenSSL
         // returns 0 if the session is NULL, so do that here.
+        // TODO(nmittler): should we throw here?
```
Comment (normanmaurer) [inline src/main/c/ssl.c]: nope we should not throw

**Comment by normanmaurer** [comment]
@nmittler one comment.

File: src/main/c/ssl.c (hunk 1 above)
Comment (nmittler) [inline src/main/c/ssl.c]: @normanmaurer Just to be clear, is your concern of using tcnative with BoringSSL and an old version of netty? With the new version of Netty, throwing would be the right thing to do, since it would expose a programming error.

Comment (normanmaurer) [inline src/main/c/ssl.c]: @nmittler nope... we still may call this method if the session is not created yet even within netty. So I think we should not throw here otherwise it will be just harder to guard

Comment (nmittler) [inline src/main/c/ssl.c]: sgtm ... done.

**Comment by nmittler** [comment]
@normanmaurer shall I cherry-pick?

**Comment by normanmaurer** [comment]
@nmittler let @Scottmitch have a look as well.

File: src/main/c/ssl.c [hunk 2]
```diff
@@ -1292,9 +1337,21 @@ TCN_IMPLEMENT_CALL(jint /* status */, SSL, writeToBIO)(TCN_STDARGS,
...
                                                        jint wlen /* sizeof(wbuf) */) {
+    BIO *b = J2P(bio, BIO *);
+    void *w = J2P(wbuf, void *);
+
+    if (b == NULL) {
+        tcn_ThrowException(e, "bio is null");
+        return 0;
+    }
+    if (w == NULL) {
+        tcn_ThrowException(e, "buffer is null");
```
Comment (normanmaurer) [inline src/main/c/ssl.c]: wbuf is null ?

File: src/main/c/ssl.c [hunk 3]
```diff
@@ -1303,9 +1360,21 @@ TCN_IMPLEMENT_CALL(jint /* status */, SSL, readFromBIO)(TCN_STDARGS,
...
                                                         jint rlen /* sizeof(rbuf) - 1 */) {
+    BIO *b = J2P(bio, BIO *);
+    void *r = J2P(rbuf, void *);
+
+    if (b == NULL) {
+        tcn_ThrowException(e, "bio is null");
+        return 0;
+    }
+    if (r == NULL) {
+        tcn_ThrowException(e, "buffer is null");
```
Comment (normanmaurer) [inline src/main/c/ssl.c]: rbuf is null ?

File: src/main/c/ssl.c [hunk 4]
```diff
@@ -1323,32 +1404,62 @@ TCN_IMPLEMENT_CALL(jint /* status */, SSL, readFromSSL)(TCN_STDARGS,
...
                                                         jint rlen /* sizeof(rbuf) - 1 */) {
+    SSL *ssl_ = J2P(ssl, SSL *);
+    void *r = J2P(rbuf, void *);
+
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return 0;
+    }
+    if (r == NULL) {
+        tcn_ThrowException(e, "buffer is null");
```
Comment (normanmaurer) [inline src/main/c/ssl.c]: rbuf is null

File: src/main/c/ssl.c (hunk 2 above)
Comment (nmittler) [inline src/main/c/ssl.c]: done.

File: src/main/c/ssl.c (hunk 3 above)
Comment (nmittler) [inline src/main/c/ssl.c]: done.

File: src/main/c/ssl.c (hunk 4 above)
Comment (nmittler) [inline src/main/c/ssl.c]: done.

**Comment by nmittler** [comment]
@Scottmitch do you want to have a look?

**Comment by Scottmitch** [comment]
@nmittler - Will look this afternoon.

File: src/main/c/ssl.c [hunk 5]
```diff
@@ -1225,9 +1230,17 @@ TCN_IMPLEMENT_CALL(jlong /* SSL * */, SSL, newSSL)(TCN_STDARGS,
     int *handshakeCount = malloc(sizeof(int));
```
Comment (Scottmitch) [inline src/main/c/ssl.c]: potential memory leak. do this after you check pre conditions.

File: src/main/c/ssl.c [hunk 6]
```diff
@@ -1366,59 +1477,82 @@ TCN_IMPLEMENT_CALL(jlong, SSL, makeNetworkBIO)(TCN_STDARGS,
...
- fail:
-    return 0;
 }
 
 // Free a BIO * (typically, the network BIO)
 TCN_IMPLEMENT_CALL(void, SSL, freeBIO)(TCN_STDARGS,
                                        jlong bio /* BIO * */) {
-    BIO *bio_;
+    BIO *bio_ = J2P(bio, BIO *);
+    if (bio_ == NULL) {
```
Comment (Scottmitch) [inline src/main/c/ssl.c]: nit: New line consistency? other functions have a new line in between declarations/initialization and preconditions.  I would prefer no-newline...but just as long as we are consistent.

File: src/main/c/ssl.c [hunk 7]
```diff
@@ -1752,14 +1887,13 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
     SSL *ssl_ = J2P(ssl, SSL *);
     TCN_ALLOC_CSTRING(ciphers);
```
Comment (Scottmitch) [inline src/main/c/ssl.c]: potential memory leak? move until after pre condition check.

**Comment by Scottmitch** [comment]
@nmittler - code looks good.

Does Netty ever call these functions where it is OK to not throw exceptions? I guess you will say those are netty bugs and we should fix them there ... but then do we need to check in both places? Seems like in some instances we will have 3 layers all checking the same thing (netty, tcnative, openssl).

**Comment by normanmaurer** [comment]
The only place where this can happen (in terms of netty) is where we use ssl_->session. The session may be NULL

File: src/main/c/ssl.c (hunk 5 above)
Comment (nmittler) [inline src/main/c/ssl.c]: This seems unrelated to this PR ... or am I missing something?

Comment (nmittler) [inline src/main/c/ssl.c]: Oh nevermind ... haven't had my coffee yet :)

Comment (nmittler) [inline src/main/c/ssl.c]: done.

File: src/main/c/ssl.c (hunk 6 above)
Comment (nmittler) [inline src/main/c/ssl.c]: done.

File: src/main/c/ssl.c (hunk 7 above)
Comment (nmittler) [inline src/main/c/ssl.c]: done.

**Comment by nmittler** [comment]
@Scottmitch PTAL

File: src/main/c/ssl.c [hunk 8]
```diff
@@ -1260,31 +1272,63 @@ TCN_IMPLEMENT_CALL(void, SSL, setBIO)(TCN_STDARGS,
...
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return;
+    }
+
     UNREFERENCED_STDARGS;
-    SSL_set_bio(J2P(ssl, SSL *), J2P(rbio, BIO *), J2P(wbio, BIO *));
+
+    SSL_set_bio(ssl_, r, w);
     return;
```
Comment (Scottmitch) [inline src/main/c/ssl.c]: cleanup: remove?

File: src/main/c/ssl.c [hunk 9]
```diff
@@ -1366,59 +1476,83 @@ TCN_IMPLEMENT_CALL(jlong, SSL, makeNetworkBIO)(TCN_STDARGS,
...
 }
 
 // Free a BIO * (typically, the network BIO)
 TCN_IMPLEMENT_CALL(void, SSL, freeBIO)(TCN_STDARGS,
                                        jlong bio /* BIO * */) {
-    BIO *bio_;
+    BIO *bio_ = J2P(bio, BIO *);
+
+    if (bio_ == NULL) {
+        tcn_ThrowException(e, "bio is null");
```
Comment (Scottmitch) [inline src/main/c/ssl.c]: For `closeBIO` we didn't throw an exception but instead we just did the close if the pointer was not null. Should we do something similar here?

File: src/main/c/ssl.c [hunk 10]
```diff
@@ -1366,59 +1476,83 @@ TCN_IMPLEMENT_CALL(jlong, SSL, makeNetworkBIO)(TCN_STDARGS,
...
 
 // Read which cipher was negotiated for the given SSL *.
 TCN_IMPLEMENT_CALL(jstring, SSL, getCipherForSSL)(TCN_STDARGS,
                                                   jlong ssl /* SSL * */)
 {
+    SSL *ssl_ = J2P(ssl, SSL *);
+
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return AJP_TO_JSTRING("");
```
Comment (Scottmitch) [inline src/main/c/ssl.c]: not sure if signatures lines up ... but can we just return null here? We shouldn't get to this line anyways right?

**Comment by Scottmitch** [comment]
@nmittler - just a few questions. LGTM after these are worked out.

File: src/main/c/ssl.c (hunk 8 above)
Comment (nmittler) [inline src/main/c/ssl.c]: done.

File: src/main/c/ssl.c (hunk 9 above)
Comment (nmittler) [inline src/main/c/ssl.c]: done.

File: src/main/c/ssl.c (hunk 10 above)
Comment (nmittler) [inline src/main/c/ssl.c]: @normanmaurer is `NULL` ok here?

Comment (normanmaurer) [inline src/main/c/ssl.c]: @nmittler NULL is ok as we throw the exception anyway and it is just to return early.

**Comment by nmittler** [comment]
@Scottmitch @normanmaurer anything else before I cherry-pick?

**Comment by normanmaurer** [comment]
Fine by me.. @Scottmitch?

**Comment by nmittler** [comment]
Cherry-picked as 83b3036

**Comment by Scottmitch** [comment]
thanks @nmittler !

**Comment by normanmaurer** [comment]
@nmittler :heart_eyes:

**Comment by nmittler** [comment]
Ha!  I feel the love :)
